src/Microsoft.Extensions.Configuration.AzureAppConfiguration/AzureAppConfigurationProvider.cs

```diff
@@ -742,5 +756,84 @@ private void UpdateCacheExpirationTime(KeyValueWatcher changeWatcher, bool succe
 
             changeWatcher.CacheExpires = DateTimeOffset.UtcNow.Add(cacheExpirationTime);
         }
+
+        private async Task ExecuteWithFailOverPolicyAsync(Func<ConfigurationClient, Task> funcToExecute, CancellationToken cancellationToken = default)
+        {
+            using IEnumerator<ConfigurationClient> clientEnumerator = _configClientProvider.GetClients().GetEnumerator();
+
+            // Guaranteed to return at least 1 element
+            Debug.Assert(clientEnumerator.MoveNext());
+
+            while (true)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                try
+                {
+                    await funcToExecute(clientEnumerator.Current);
+
+                    _configClientProvider.UpdateClientStatus(clientEnumerator.Current, successful: true);
+
+                    break;
+                }
+                catch (AggregateException ae)

```

Comment:
We call `UpdateCacheExpirationTime` in the `finally` block after making any network call during refresh. It doesn't matter if the refresh operation failed or succeeded. 
If the operation failed, we would still push the cache expiration time forward, so the next failover client will not refresh anything since the cache hasn't expired yet.
We need to figure out a way to update cache expiration time only when:
- refresh operation is successful; OR
- when all failover clients have had a chance to refresh configuration and all have failed.

Reply:
Good catch! Thanks!
I have now added a function `shouldUpdateCacheExpirationOnFailure` which basically tells if the current client is the last client in the client list and if we need to update the cache expiration time. 
With the new logic, 
1. We always update the cache expiration time if the request was successful. OR
2. The request was failed and all clients have been exhausted.

Reply:
I updated the logic even further. The success case is handled in the `LoadAll` method as earlier, but the failure case is now handled at the caller level when all of the clients are exhausted.